src/merkle/simple_smt/mod.rs

```diff
@@ -66,36 +60,41 @@ impl SimpleSmt {
         })
     }
 
-    /// Appends the provided entries as leaves of the tree.
+    /// Returns a new [SimpleSmt] instantiated with the specified depth and with leaves
+    /// set as specified by the provided entries.
     ///
-    /// # Errors
+    /// All leaves omitted from the entries list are set to [ZERO; 4].
     ///
-    /// The function will fail if the provided entries count exceed the maximum tree capacity, that
-    /// is `2^{depth}`.
-    pub fn with_leaves<R, I>(mut self, entries: R) -> Result<Self, MerkleError>
+    /// # Errors
+    /// Returns an error if:
+    /// - If the depth is 0 or is greater than 64.
+    /// - The number of entries exceeds the maximum tree capacity, that is 2^{depth}.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_leaves<R, I>(depth: u8, entries: R) -> Result<Self, MerkleError>
     where
         R: IntoIterator<IntoIter = I>,
         I: Iterator<Item = (u64, Word)> + ExactSizeIterator,
     {
-        // check if the leaves count will fit the depth setup
-        let mut entries = entries.into_iter();
-        let max = 1 << self.depth.min(63);
+        // create an empty tree
+        let mut tree = Self::new(depth)?;
+
+        // check if the number of leaves can be accommodated by the tree's depth; we use a min
+        // depth of 63 because we consider passing in a vector of size 2^64 infeasible.
+        let entries = entries.into_iter();
+        let max = 1 << tree.depth.min(63);
         if entries.len() > max {
-            return Err(MerkleError::InvalidEntriesCount(max, entries.len()));
+            return Err(MerkleError::InvalidNumEntries(max, entries.len()));
         }
 
-        // append leaves and return
-        entries.try_for_each(|(key, leaf)| self.insert_leaf(key, leaf))?;
-        Ok(self)
-    }
-
-    /// Replaces the internal empty digests used when a given depth doesn't contain a node.
-    pub fn with_empty_subtrees<I>(mut self, hashes: I) -> Self
-    where
-        I: IntoIterator<Item = RpoDigest>,
-    {
-        self.replace_empty_subtrees(hashes.into_iter().collect());
-        self
+        // append leaves to the tree returning an error if a duplicate entry for the same key
+        // is found
+        for (key, value) in entries {
+            let old_value = tree.update_leaf(key, value)?;
+            if old_value != EMPTY_WORD {
+                return Err(MerkleError::DuplicateValuesForKey(key));
```

Comment:
An edge case that wouldn't be caught here is:

```rust
let entries = [(10, EMPTY_WORD), (10, [ONE, ONE, ONE, ONE])];
```

Should we be concerned about this?

Reply:
Good point! Fixed!

Reply:
I think this will error whenever we insert a node with an empty word even if we don't have duplicates.  This condition `empty_entries.contains(&key)` will be triggered due to the insert on line 93 `empty_entries.insert(key)`.  A test case that should work but currently fails:

```rust
#[test]
fn fail_no_duplicates_empty_node() {
    let entries = [(1_u64, int_to_node(0)), (5, int_to_node(2))];
    let smt = SimpleSmt::with_leaves(64, entries);
    assert!(smt.is_ok());
}
```

Reply:
Looks good!